sql-plugin/src/main/java/org/apache/spark/sql/catalyst/CudfUnsafeRow.java

```diff
@@ -0,0 +1,416 @@
+   *
+   * @param attributes the schema of what this will hold.  This is the schema of the underlying
+   *                    row, so if columns were re-ordered it is the attributes of the reordered
+   *                    data.
+   * @param remapping a mapping from the user requested column to the underlying column in the
+   *                  backing row.
+   */
+  public CudfUnsafeRow(Attribute[] attributes, int[] remapping) {
+    int offset = 0;
+    startOffsets = new int[attributes.length];
+    for (int i = 0; i < attributes.length; i++) {
+      Attribute attr = attributes[i];
+      int length = GpuColumnVector.getRapidsType(attr.dataType()).getSizeInBytes();
+      offset = alignOffset(offset, length);
+      startOffsets[i] = offset;
+      offset += length;
+    }
+    fixedWidthSizeInBytes = offset;
+    this.remapping = remapping;
+    assert startOffsets.length == remapping.length;
+  }
+
+  // for serializer
+  public CudfUnsafeRow() {}
+
+  @Override
+  public int numFields() { return startOffsets.length; }
+
+  /**
+   * Update this CudfUnsafeRow to point to different backing data.
+   *
+   * @param address the address in host memory for this.  We should change this to be a
+   *                MemoryBuffer class or something like that.
+   * @param sizeInBytes the size of this row's backing data, in bytes
+   */
+  public void pointTo(long address, int sizeInBytes) {
+    assert startOffsets != null && startOffsets.length > 0 : "startOffsets not properly initialized";
+    assert sizeInBytes % 8 == 0 : "sizeInBytes (" + sizeInBytes + ") should be a multiple of 8";
+    this.address = address;
+    this.sizeInBytes = sizeInBytes;
+  }
+
+  @Override
+  public void update(int ordinal, Object value) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public Object get(int ordinal, DataType dataType) {
+    // Don't remap the ordinal because it will be remapped in each of the other backing APIs
+    return SpecializedGettersReader.read(this, ordinal, dataType, true, true);
+  }
+
+  @Override
+  public boolean isNullAt(int ordinal) {
+    int i = remapping[ordinal];
+    assertIndexIsValid(i);
+    int validByteIndex = i / 8;
+    int validBitIndex = i % 8;
+    byte b = Platform.getByte(null, address + fixedWidthSizeInBytes + validByteIndex);
+    return ((1 << validBitIndex) & b) == 0;
+  }
+
+  @Override
+  public void setNullAt(int ordinal) {
+    int i = remapping[ordinal];
+    assertIndexIsValid(i);
+    int validByteIndex = i / 8;
+    int validBitIndex = i % 8;
+    byte b = Platform.getByte(null, address + fixedWidthSizeInBytes + validByteIndex);
+    b = (byte)((b & ~(1 << validBitIndex)) & 0xFF);
+    Platform.putByte(null, address + fixedWidthSizeInBytes + validByteIndex, b);
+  }
+
+  @Override
+  public boolean getBoolean(int ordinal) {
+    int i = remapping[ordinal];
+    assertIndexIsValid(i);
+    return Platform.getBoolean(null, getFieldAddress(i));
```

Comment:
Worth having a `getFieldAddressFromOrdinal` to reduce some of the boilerplate in these methods?